src/tracing/mod.rs

```diff
@@ -27,6 +27,169 @@ use crate::eth_provider::{
 
 pub type TracerResult<T> = Result<T, EthApiError>;
 
+/// Represents the result of tracing a transaction.
+type TracingStateResult = TracerResult<(TracingResult, reth_revm::primitives::State)>;
+
+/// Representing different tracing options for transactions.
+#[derive(Clone, Debug)]
+enum TracingOptions {
+    /// Geth debug tracing options.
+    Geth(GethDebugTracingOptions),
+    /// Parity tracing options.
+    Parity(TracingInspectorConfig),
+}
+
+/// Representing the result of tracing transactions.
+#[derive(Clone, Debug)]
+enum TracingResult {
+    /// Geth trace results.
+    Geth(Vec<TraceResult>),
+    /// Parity trace results.
+    Parity(Vec<LocalizedTransactionTrace>),
+}
+
+impl TracingResult {
+    /// Converts the tracing result into Geth traces.
+    fn into_geth(self) -> Option<Vec<TraceResult>> {
+        if let Self::Geth(traces) = self {
+            Some(traces)
+        } else {
+            None
+        }
+    }
+
+    /// Converts the tracing result into Parity traces.
+    fn into_parity(self) -> Option<Vec<LocalizedTransactionTrace>> {
+        if let Self::Parity(traces) = self {
+            Some(traces)
+        } else {
+            None
+        }
+    }
+
+    /// Returns a closure that transacts and retrieves the traces for the given transaction.
+    /// Captures the `opts` to use within the closure.
+    fn trace<P: EthereumProvider + Send + Sync + Clone>(
+        opts: TracingOptions,
+    ) -> impl Fn(EnvWithHandlerCfg, &mut EthDatabaseSnapshot<P>, &reth_rpc_types::Transaction) -> TracingStateResult
+    {
+        move |env: EnvWithHandlerCfg,
+              db: &mut EthDatabaseSnapshot<P>,
+              tx: &reth_rpc_types::Transaction|
+              -> TracingStateResult { Self::trace_transaction(env, db, tx, opts.clone()) }
+    }
+
+    /// Traces the transaction based on the specified tracing options and returns the resulting traces and state.
+    fn trace_transaction<P: EthereumProvider + Send + Sync + Clone>(
+        env: EnvWithHandlerCfg,
+        db: &mut EthDatabaseSnapshot<P>,
+        tx: &reth_rpc_types::Transaction,
+        tracing_options: TracingOptions,
+    ) -> TracingStateResult {
+        match tracing_options {
+            TracingOptions::Geth(opts) => Self::trace_geth(env, db, tx, opts),
+            TracingOptions::Parity(tracing_config) => Self::trace_parity(env, db, tx, tracing_config),
+        }
+    }
+
+    /// Traces the transaction with Geth tracing options and returns the resulting traces and state.
+    fn trace_geth<P: EthereumProvider + Send + Sync + Clone>(
```

Comment:
I'm just wondering if it makes sense to have `trace_geth, `trace_parity` and `trace_transaction` implemented on `TracingResult`, since it's not really linked to the result and all three of the function are static so they don't event need to use the result. I would maybe move them to pure function or implement them on the tracer.

Reply:
I've just simplified a lot the logic by simply encapsulating our `tracing_options` that are

```rust
/// Representing different tracing options for transactions.
#[derive(Clone, Debug)]
pub enum TracingOptions {
    /// Geth debug tracing options.
    Geth(GethDebugTracingOptions),
    /// Parity tracing options.
    Parity(TracingInspectorConfig),
}
```

directly inside our `Tracer` in order to be able to play with it in an easy way. Let me know what you think about this?